XWIKI-10252: Add and/or edit objects in a page

woshilapin opened the pull request:
* Provide a new object policy to edit but also create new objects in on step

tmortagne commented:
+1 for inclusion if the only thing against it is JMock

woshilapin commented:
Do I need to do something more (except for the JMock/Mockito problem)?

vmassol commented:
There were some comments to which you didn't answer AFAICS woshilapin

vmassol commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -3447,11 +3459,153 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+     * @param eform
+     * @param context
+     * @throws XWikiException
+     */
```
missing a @since annotation

woshilapin replied:
OK, which version should I put?

vmassol commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/EditForm.java:
```diff
@@ -280,4 +282,16 @@ public void setHidden(String hidden)
         this.hidden = hidden;
     }
 
+    public ObjectPolicyType getObjectPolicy() {
```
missing javadoc + @since annotation

woshilapin replied:
Same as before, what to add for @since?
I added the javadoc.

vmassol commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ObjectPolicyType.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Enumeration of supported object policy types.
+ * 
+ * @version $Id$
```
missing a @since annotation. Also missing an @Unstable annotation since you're putting this as a public API

woshilapin replied:
I've added a @Unstable here.  However, can you explain me why this is a public API here?  I was thinking that the public API is in com.xpn.xwiki.api package. Public for who? (I'm asking for my knowledge here!).

vmassol commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/ObjectPolicyType.java:
```diff
@@ -0,0 +1,65 @@
+package com.xpn.xwiki.web;
+
+/**
+ * Enumeration of supported object policy types.
```
What is an object policy type? It should be explained. Otherwise this javadoc is not very useful :)

woshilapin replied:
Done.  Not sure however that everything I put here should be put at this place.

vmassol commented on xwiki-platform-core/xwiki-platform-oldcore/src/test/java/com/xpn/xwiki/doc/XWikiDocumentTest.java:
```diff
@@ -34,8 +34,12 @@
 import org.apache.velocity.VelocityContext;
+import org.jmock.Expectations;
 import org.jmock.Mock;
+import org.jmock.Mockery;
```
Yes I agree that we need to update this to Mockito. It's easy to do.

vmassol replied:
How much is covered of the new code? There doesn't seem to be many tests compared to all the code added? I'm worried about that since we're touching the core here, we need to be careful not to introduce bugs.

vmassol commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -3447,11 +3459,153 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+     *     }
+     *   ]
+     * }
+     * 
```
you need to check your javadoc. It won't be rendered properly you need to use "code" and "pre" html tags for example. Try generating the javadoc (or just press ctrl+J in IDEA for example) to see what I mean.

woshilapin replied:
Should be OK now.

vmassol commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -3447,11 +3459,153 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+     * }
+     * 
+     * @param request The input HTTP request that provides the parameters
+     * @return The map containing ordered data (see more details in the method's documentation)
```
not sure it's a good idea to refer to the implementation for more details or we could say that everywhere :) I'd rather you remove this text

woshilapin replied:
Done.

vmassol commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -3447,11 +3459,153 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+     * * XWiki.XWikiRights_0_users=XWiki.Me
+     * * XWiki.XWikiRights_0_groups=XWiki.XWikiAllGroup
+     * * XWiki.XWikiRights_1_user=XWiki.Admin
+     * * XWiki.XWikiUsers_1_name=Spirou
```
For a list you need to use ````<ul><li>````

woshilapin replied:
Done.

vmassol commented:
Woshi, if you want we can work together on your pull request on Wednesday, when I'm in Paris. I can show you how to move it to Mockito and we'll fix the small details too.

sdumitriu commented:
I would **really** prefer to have the more powerful approach I proposed instead of this one.

woshilapin commented:
@vmassol, thanks for the proposition and for your comments.  I'll try to address the comments before wednesday and yes, it would be nice to have help on Mockito stuff.

@sdumitriu, I'd like too but I have no time for it (it's a more complex one).  However, my proposition (which can create Objects) will live next to the default behavior of XWiki (which doesn't create Objects); it's for backward compatibility reasons.  See the `if` (which could be a `switch`) https://github.com/xwiki/xwiki-platform/pull/271/files#diff-9c8228109f081a40ad73bfd95c485902R3604, you'll see that my proposition is enabled only with the `objectPolicy=updateOrCreate` parameter in the query string.  If this parameter is not given, then the system fallback to the old policy which is not creating Objects.  It's then easy to create a third implementation which will be enabled with `objectPolicy=CRUD` (if we call this new implementation `CRUD`).  Hoep my explanations make sense.

woshilapin commented:
I tried to provide more documentation.  Still don't know which value I should put for @since.  And waiting for Mockito stuff tomorrow.

woshilapin commented:
In the last commit, I did a few new things.

* First, I merged with last version of 'master' branch (should ease the work of the merger!).
* I added the @since with 7.0RC1 as discussed with @vmassol 
* I removed the tests from `XWikiDocumentTest.java` to move them to `XWikiDocumentMockitoTest.java`

But then, I still have some incomplete stuff.

First of all, at the time of the first version of the Pull request, I added 2 integration tests, one that concerns directly the code that I'm adding (`readObjectsFromFormUpdateOrCreate`) and another one for the already existing method (`readObjectsFromForm`).  When I moved everything to Mockito, my stuff works pretty well but I still have problem for the other one (seems that there is some missing mock into `<BaseObject>.fromMap()̀` but I was not able to make it work).

The second thing I'm missing is to check the messages from the logger. Not sure if I need to do that into another integration test function or if I can do it my big integration test? And how to do it (I've saw some stuff about that in file like `PropertyConverterTest.java` but not sure I can do the same since we have the old core Rules).

woshilapin commented:
It seems that I forgot a bit of the discussion I had with @vmassol about the logger problem.  He just told me (again!) that testing the log messages would imply more important modifications on the `XWikiDocument` class and a more custom testing class.  We decided to not do it for this PR (it may still be implemented later).

So the only thing who's missing is the problem with `<BaseObject>.fromMap()`.  We'll try to investigate that with @vmassol today.

woshilapin commented:
OK, I fixed the problem and applied the coding style from XWiki.  Should be OK for reviewing now.

Enygma2002 commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -252,6 +256,13 @@ public String toString()
+     * {@code &lt;space&gt;.&lt;pageClass&gt;_&lt;number&gt;_&lt;propertyName&gt;}
+     * (e.g. {@code XWiki.XWikiRights_0_member}).
+     */
+    private static final Pattern XPROPERTY_URL_REFERENCE = Pattern.compile("(.+?)_([0-9]+)_(.+)");
```
Should probably be named `XPROPERTY_REFERENCE_PATTERN`, for clarity, since the variable is not a EntityReference.

woshilapin replied:
Name has been modified.

Enygma2002 commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -3446,11 +3457,150 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+            String classPropertyName;
+            className = matcher.group(1);
+            classNumberAsString = matcher.group(2);
+            classPropertyName = matcher.group(3);
```
No need to separate declaration from initialization for these variables, i.e. just use:
`String className = matcher.group(1);`

...except for `classNumber`, of course, which is initialized in the try block below.

woshilapin replied:
Factorization has been done.

Enygma2002 commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -3446,11 +3457,150 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+            }
+            // Get the property from the right object #objectNumber of type 'objectName'; create it if they don't exist
+            Map<String, String[]> object = objectMap.get(classNumber);
+            if (objectMap.get(classNumber) == null) {
```
You should reuse `object` instead of re-evaluating `objectMap.get(classNumber)`.

woshilapin replied:
Yes, I changed it.

Enygma2002 commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -3446,11 +3457,150 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+            className = matcher.group(1);
+            classNumberAsString = matcher.group(2);
+            classPropertyName = matcher.group(3);
+            DocumentReference classReference = getCurrentDocumentReferenceResolver().resolve(className);
```
Why not use the DocumentReference as a map key for `result`? Both there and below, in `readObjectsFromFormUpdateOrCreate` you end up resolving the className to a DocumentReference.

Said differently:
If you use the `classReference` resolved here as key for `result` you can avoid re-resolving below, in `readObjectsFromFormUpdateOrCreate` and use the key directly.

woshilapin replied:
Indeed, good point. I've done the modification.

Enygma2002 commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -3446,11 +3457,150 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+            parseRequestUpdateOrCreate(eform.getRequest(), context);
+        for (Entry<String, SortedMap<Integer, Map<String, String[]>>> requestClassEntries : fromRequest.entrySet()) {
+            DocumentReference requestClassReference =
+                getCurrentDocumentReferenceResolver().resolve(requestClassEntries.getKey());
```
As mentioned above, if your key is a DocumentRefence, you don`t need to resolve again the same class name string.

woshilapin replied:
See above, modification has been done.

Enygma2002 commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -3446,11 +3457,150 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+                newObject.setOwnerDocument(this);
+                newObjects.add(newObject);
+            }
+            getXObjects().put(requestClassReference, newObjects);
```
Doesn't this override **all** the objects of type `requestClassReference` in the document with **only** the documents from the current request? Existing objects not modified by this request should be preserved.

Also, using the internal list directly, instead of going though XWikiDocument's add/set/remove methods might introduce inconsistencies wrt the expectations on the stored list of objects that these methods seem to take a decent amount of care to enforce, i.e. an object must be set in the list of objects for the class on the same index as it's object number, etc.

IMO, getXObjects() should never return the internal objects list, but a copy. However, this is a different topic.

woshilapin replied:
[`java.util.Map.put()`](https://docs.oracle.com/javase/7/docs/api/java/util/Map.html#put%28K,%20V%29), which is the object returned by `getXObjects()` method, doesn't replace content for different keys.  In this case, other classes contains other keys so they should not be destroyed.

Enygma2002 replied:
But it does replace all the content for the "put" key, right? In this case, the key being placed is `requestClassReference` as I have mentioned, however `newObjects` contains only objects of type `requestClassReference` that are present in **the request** (i.e. updated or created objects only). `newObjects` does not consider the objects that were **already present** in the document, so any such objects will be **removed** and only the `newObjects` will be stored. Correct?

woshilapin replied:
It seems not.  I added some additional assertions in my tests to test this case (see the assert on `int` on object at index 0 and the assert on `string` on object at index 1).

Enygma2002 replied:
You miss my point and the details I offered in the comment on `generateFakeObjects`.

You are adding 2 objects in your test's setup (index 0 and 1), in other words, these 2 objects already exist in your tested document. Your actual test's execution (`generateFakeObjects`) edits these 2 objects index 0, 1 and creates object index 2 and also tries to create object index 42 (which is not a consecutive index and fails).

In your above test you miss the case when your test's setup's document contains 3 objects, 2 of which are updated by the test's execution (index 0 and 1) and a 3rd object that should not be touched at all by your test's execution and should still be there once the test ends.

Example:
Given: index 0, index 1, index 2
Execute: update (index 0, index 1), create (index 3, index 42)
Assert: updated (index 0, index 1), unchanged/still existing (index 2), created (index 3), not created (index 42)

woshilapin replied:
Indeed, it spotted a bug.  I added the missing tests and fixed the bug.

Enygma2002 commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/EditForm.java:
```diff
@@ -280,4 +282,34 @@ public void setHidden(String hidden)
+	 *
+	 * @return Return the Object Policy type
+	 * @since 7.0RC1
+	 */
```
Bad javadoc indentation.

woshilapin replied:
Not sure what is bad.  I applied the default one in Eclipse (with `codestyle.xml` imported), I hope it's fine now.

vmassol commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/EditForm.java:
```diff
@@ -280,4 +282,34 @@ public void setHidden(String hidden)
+	 * {@link com.xpn.xwiki.web.ObjectPolicyType ObjectPolicyType} for more
+	 * information about what is an object policy.
+	 *
+	 * @return Return the Object Policy type
```
actually it should be (without uppercase):

     @return the Object Policy type

woshilapin replied:
Should be OK now.

Enygma2002 commented on xwiki-platform-core/xwiki-platform-oldcore/src/test/java/com/xpn/xwiki/doc/XWikiDocumentMockitoTest.java:
```diff
@@ -145,4 +163,169 @@ public void setTranslationAsContextDoc() throws Exception
+        try {
+            baseObject = this.document.newXObject(this.document.getDocumentReference(), this.oldcore.getXWikiContext());
+            baseObject2 =
+                this.document.newXObject(this.document.getDocumentReference(), this.oldcore.getXWikiContext());
```
You probably need to add more objects (right now you have index 0 and 1 already existing and overridden by the request) so that you get to index 3 for example, that the request does not touch and you need to ensure that this object is still there after the request is processed (i.e. object with index 3 that was previously existing is not removed by the request processing).

...basically test the case I mentioned above when you were doing ` getXObjects().put(requestClassReference, newObjects);`

Enygma2002 commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -3446,11 +3457,150 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+                Map<String, String[]> requestObjectPropertyMap = requestObjectEntry.getValue();
+                BaseObject oldObject = getXObject(requestClassReference, requestObjectNumber);
+                if (oldObject == null) {
+                    // Create the object only if it has been numbered one more than the number of existing objects
```
Not entirely relevant to this particular issue, since what you want to do now is update and/or create, but what about delete?

We currently have a separate action for deleting objects AFAIK, but if we are updating and creating in the same request, why not delete as well?

Basically, the case would be `oldObject == null && (requestObjectPropertyMap == null || requestObjectPropertyMap.isEmpty())`, signaling that the requests explicitly states that the object with a particular number is to be deleted, probably something like `space.className_objNumber=<emptyString>`.

Of course, this would need changing the parsing as well and is probably for a separate issue to be handled, if ever, and is obviously opened to discussion since some people might consider it error prone.

woshilapin replied:
I would prefer to not put it in this PR and leave it for another issue as you suggest.  And by the way, a more complete proposition has also been done on the Developer Mailing List (see [[xwiki-devs] [Proposal] Automatic creation of objects with HTTP parameters](https://www.mail-archive.com/devs%40xwiki.org/msg27580.html)).

Enygma2002 commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -3446,11 +3457,147 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+     * }
+     * </pre></code>
+     * 
+     * @param request The input HTTP request that provides the parameters
```
As @vmassol mentioned, all javadoc (this case and all below) should start with lower case for @params / @returns / etc.; except for method/class descriptions which are regular sentences.

woshilapin replied:
Oops, sorry, indeed, I corrected only the one that was mentioned.

Enygma2002 commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -3446,11 +3457,148 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+                if (oldObject == null) {
+                    // Create the object only if it has been numbered one more than the number of existing objects
+                    if (requestObjectPropertyMap != null
+                        && requestObjectNumber == getXObjectSize(requestClassReference)) {
```
BTW, why are we not allowing client-chosen object IDs and dropping these requests? Asking since we are expecting the client to choose the ID and AFAIK there is no proper way for clients to determine what is the next ID in the sequence and "holes" in object number sequences are accepted in our model.

woshilapin replied:
My choice is based on the answer to this question: "what will you do if there is no object in the wiki page and your client gives you something like `XWiki.XWikiRights_42_levels`?".  You should create 42 objects only for one useful object?  That's the reason I made this choice.  I kind of force the developer to know what he's doing.  It's only a limitation, may be improved in a later PR?

Enygma2002 replied:
Actually, you do not create 41 objects just to add one. As far as I can see in the API, it's just a matter of creating a list filled with 41 `null` values and set your object on the 42nd position. See [XWikiDocument.setXObject(int nb, BaseObject object)](https://github.com/xwiki/xwiki-platform/blob/1769a8b2b55369b9faf5d08628760085c22e96a0/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java#L2642)

I too find the implementation a bit peculiar, but that still remains an implementation details. As long as the current java API supports this, I see no reason not to support it inside this action. 

Don`t you agree?

cjdelisle commented:
Subject: A bike shed (any colour will do) on greener grass...
From: Poul-Henning Kamp <phk@freebsd.org>
Date: Sat, 02 Oct 1999 16:14:10 +0200
Message-ID: <18238.938873650@critter.freebsd.dk>
Sender: phk@critter.freebsd.dk
Bcc: Blind Distribution List: ;
MIME-Version: 1.0


[bcc'ed to committers, hackers]

My last pamphlet was sufficiently well received that I was not
scared away from sending another one, and today I have the time
and inclination to do so.

I've had a little trouble with deciding on the right distribution
of this kind of stuff, this time it is bcc'ed to committers and
hackers, that is probably the best I can do.  I'm not subscribed
to hackers myself but more on that later.

The thing which have triggered me this time is the "sleep(1) should
do fractional seconds" thread, which have pestered our lives for
many days now, it's probably already a couple of weeks, I can't
even be bothered to check.

To those of you who have missed this particular thread: Congratulations.

It was a proposal to make sleep(1) DTRT if given a non-integer
argument that set this particular grass-fire off.  I'm not going
to say anymore about it than that, because it is a much smaller
item than one would expect from the length of the thread, and it
has already received far more attention than some of the *problems*
we have around here.

The sleep(1) saga is the most blatant example of a bike shed
discussion we have had ever in FreeBSD.  The proposal was well
thought out, we would gain compatibility with OpenBSD and NetBSD,
and still be fully compatible with any code anyone ever wrote.

Yet so many objections, proposals and changes were raised and
launched that one would think the change would have plugged all
the holes in swiss cheese or changed the taste of Coca Cola or
something similar serious.

"What is it about this bike shed ?" Some of you have asked me.

It's a long story, or rather it's an old story, but it is quite
short actually.  C. Northcote Parkinson wrote a book in the early
1960'ies, called "Parkinson's Law", which contains a lot of insight
into the dynamics of management.

You can find it on Amazon, and maybe also in your dads book-shelf,
it is well worth its price and the time to read it either way,
if you like Dilbert, you'll like Parkinson.

Somebody recently told me that he had read it and found that only
about 50% of it applied these days.  That is pretty darn good I
would say, many of the modern management books have hit-rates a
lot lower than that, and this one is 35+ years old.

In the specific example involving the bike shed, the other vital
component is an atomic power-plant, I guess that illustrates the
age of the book.

Parkinson shows how you can go in to the board of directors and
get approval for building a multi-million or even billion dollar
atomic power plant, but if you want to build a bike shed you will
be tangled up in endless discussions.

Parkinson explains that this is because an atomic plant is so vast,
so expensive and so complicated that people cannot grasp it, and
rather than try, they fall back on the assumption that somebody
else checked all the details before it got this far.   Richard P.
Feynmann gives a couple of interesting, and very much to the point,
examples relating to Los Alamos in his books.

A bike shed on the other hand.  Anyone can build one of those over
a weekend, and still have time to watch the game on TV.  So no
matter how well prepared, no matter how reasonable you are with
your proposal, somebody will seize the chance to show that he is
doing his job, that he is paying attention, that he is *here*.

In Denmark we call it "setting your fingerprint".  It is about
personal pride and prestige, it is about being able to point
somewhere and say "There!  *I* did that."  It is a strong trait in
politicians, but present in most people given the chance.  Just
think about footsteps in wet cement.

I bow my head in respect to the original proposer because he stuck
to his guns through this carpet blanking from the peanut gallery,
and the change is in our tree today.  I would have turned my back
and walked away after less than a handful of messages in that
thread.

And that brings me, as I promised earlier, to why I am not subscribed
to -hackers:

I un-subscribed from -hackers several years ago, because I could
not keep up with the email load.  Since then I have dropped off
several other lists as well for the very same reason.

And I still get a lot of email.  A lot of it gets routed to /dev/null
by filters:  People like Brett Glass will never make it onto my
screen, commits to documents in languages I don't understand
likewise, commits to ports as such.  All these things and more go
the winter way without me ever even knowing about it.

But despite these sharp teeth under my mailbox I still get too much
email.

This is where the greener grass comes into the picture:

I wish we could reduce the amount of noise in our lists and I wish
we could let people build a bike shed every so often, and I don't
really care what colour they paint it.

The first of these wishes is about being civil, sensitive and 
intelligent in our use of email.

If I could concisely and precisely define a set of criteria for
when one should and when one should not reply to an email so that
everybody would agree and abide by it, I would be a happy man, but
I am too wise to even attempt that.

But let me suggest a few pop-up windows I would like to see
mail-programs implement whenever people send or reply to email
to the lists they want me to subscribe to:

      +------------------------------------------------------------+
      | Your email is about to be sent to several hundred thousand |
      | people, who will have to spend at least 10 seconds reading |
      | it before they can decide if it is interesting.  At least  |
      | two man-weeks will be spent reading your email.  Many of   |
      | the recipients will have to pay to download your email.    |
      |								   |
      | Are you absolutely sure that your email is of sufficient   |
      | importance to bother all these people ?                    |
      |								   |
      |                  [YES]  [REVISE]  [CANCEL]                 |
      +------------------------------------------------------------+

      +------------------------------------------------------------+
      |	Warning:  You have not read all emails in this thread yet. |
      |	Somebody else may already have said what you are about to  |
      |	say in your reply.  Please read the entire thread before   |
      |	replying to any email in it.                               |
      |								   |
      |			     [CANCEL]                              |
      +------------------------------------------------------------+

      +------------------------------------------------------------+
      |	Warning:  Your mail program have not even shown you the    |
      |	entire message yet.  Logically it follows that you cannot  |
      |	possibly have read it all and understood it.               |
      |								   |
      |	It is not polite to reply to an email until you have       |
      |	read it all and thought about it.			   |
      |								   |
      | A cool off timer for this thread will prevent you from     |
      | replying to any email in this thread for the next one hour |
      |								   |
      |			      [Cancel]				   |
      +------------------------------------------------------------+

      +------------------------------------------------------------+
      | You composed this email at a rate of more than N.NN cps    |
      | It is generally not possible to think and type at a rate   |
      | faster than A.AA cps, and therefore you reply is likely to |
      | incoherent, badly thought out and/or emotional.            |
      |								   |
      | A cool off timer will prevent you from sending any email   |
      | for the next one hour.			 		   |
      |								   |
      |			      [Cancel]				   |
      +------------------------------------------------------------+

The second part of my wish is more emotional.  Obviously, the
capacities we had manning the unfriendly fire in the sleep(1)
thread, despite their many years with the project, never cared
enough to do this tiny deed, so why are they suddenly so enflamed
by somebody else so much their junior doing it ?

I wish I knew.

I do know that reasoning will have no power to stop such "reactionaire
conservatism".  It may be that these people are frustrated about
their own lack of tangible contribution lately or it may be a bad
case of "we're old and grumpy, WE know how youth should behave".

Either way it is very unproductive for the project, but I have no
suggestions for how to stop it.  The best I can suggest is to refrain
from fuelling the monsters that lurk in the mailing lists:  Ignore
them, don't answer them, forget they're there.

I hope we can get a stronger and broader base of contributors in
FreeBSD, and I hope we together can prevent the grumpy old men
and the Brett Glasses of the world from chewing them up, spitting
them out and scaring them away before they ever get a leg to the 
ground.

For the people who have been lurking out there, scared away from
participating by the gargoyles:  I can only apologise and encourage
you to try anyway, this is not the way I want the environment in
the project to be.

Poul-Henning

Enygma2002 commented:
@cjdelisle no need to get emotional or rude, for the matter. If you do not like the PR review process, feel free to discuss it on the list and propose improvements, instead of wasting time bashing (yes, sarcasm counts) people who actually take time off their regular tasks in order to review someone's PR.

At least 90% of all the issues regarding this PR have so far been purely technical.

Note, in case it's not obvious: I am being diplomatic here, but I feel offended.

woshilapin commented:
I also did a small improvement for creating the `null` objects in order to create the 42th object (as you suggest @Enygma2002.

woshilapin commented:
On top of that, I improved the documentation [URL format](http://platform.xwiki.org/xwiki/bin/view/DevGuide/URL+Format#HAction:save).  I also documented about my patch in the page but put it into a `{{comment}}` macro at the moment [(see `objectPolicy` documentation)](http://platform.xwiki.org/xwiki/bin/view/DevGuide/URL+Format#HParameter:objectPolicy).  Feel free to uncomment it when the PR is merged.

Enygma2002 commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java:
```diff
@@ -3498,11 +3509,154 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+                newObject.setOwnerDocument(this);
+                while(newObjects.size() < requestObjectNumber) {
+                    newObjects.add(null);
+                }
```
IMO, it would be nice to not expose this implementation detail (i.e. that unused object numbers are filled with `null`s inside the list of objects).

When I mentioned this to you in my previous commit, I only meant it as an explanation to what is going on in the background and reuse that existing method (instead of reproducing its behavior).

What it boils down to is the fact that, instead of managing the list of `newObjects`, with adding existing objects, adding new objects and filling gaps with `null`s and, in the end, replacing the entire list of objects for `requestClassReference` with `newObjects`, you could have avoided managing all these implementation details and you could have just used the API method I pointed out.

More precisely, at the end of this for loop, you would just say `setXObject(newObject.getNumber(), newObject)`. 
This way you properly handle both existing objects and new objects, since existing objects will be replaced with the updated version at the existing index and new objects will be set at the new index, while "holes" will be filled with `null`s by the current implementation of the API (`setXObject` method).

The result in code is that you can remove the while loop above (that adds `null`s), remove the `getXObjects().put(requestClassReference, newObjects);` below and also remove the while loop at line 3628 above (which adds existing objects to your re-managed list) and anything related to the handling of the `newObjects` list, since you don`t need to care about all those details any more.

I am going to accept this PR as it is, since you have fixed the main problem (which was the existing objects bug) and it is in a good shape now, but it would be really nice if you found the time to remove the useless exposure of technical details that I detailed above, since otherwise it makes the code more resistant to future change in the API's implementation.

woshilapin commented:
Indeed, I tried what you explained in #397 and it's a serious simplification of the code that hides these implementation details.  Thanks.  I leave you the review of this other PR.
